libdnf5/base/base.cpp

```diff
@@ -71,21 +71,77 @@ Base * Base::get_locked_base() noexcept {
     return locked_base;
 }
 
-void Base::load_defaults() {
-    const std::string file_path = LIBDNF5_DISTRIBUTION_CONFIG_FILE;
-    try {
-        log_router.debug("Loading default configuration from \"{}\"", file_path);
-        ConfigParser parser;
-        parser.read(file_path);
-        config.load_from_parser(parser, "main", vars, *get_logger(), Option::Priority::DEFAULT);
-    } catch (const libdnf5::MissingConfigError & ex) {
-        log_router.debug("Configuration file \"{}\" not found", file_path);
+void Base::load_config() {
+    ConfigParser parser;
+
+    fs::path conf_file_path{config.get_config_file_path_option().get_value()};
+    fs::path conf_dir_path{CONF_DIRECTORY};
+    fs::path distribution_conf_file_path{LIBDNF5_DISTRIBUTION_CONFIG_FILE};
+    fs::path distribution_conf_dir_path{LIBDNF5_DISTRIBUTION_CONFIG_DIR};
+
+    const auto conf_file_path_priority{config.get_config_file_path_option().get_priority()};
+    const bool use_installroot_config{
+        !config.get_use_host_config_option().get_value() && conf_file_path_priority < Option::Priority::COMMANDLINE};
+    if (use_installroot_config) {
+        fs::path installroot_path{config.get_installroot_option().get_value()};
+        conf_file_path = installroot_path / conf_file_path.relative_path();
+        conf_dir_path = installroot_path / conf_dir_path.relative_path();
+        distribution_conf_file_path = installroot_path / distribution_conf_file_path.relative_path();
+        distribution_conf_dir_path = installroot_path / distribution_conf_dir_path.relative_path();
+    }
+    const bool user_defined_config_file_name = conf_file_path_priority >= Option::Priority::COMMANDLINE;
+
+    // If the name of the user configuration file is defined or the file exists in the default location,
+    // it will be loaded instead of the distribution configuration file.
+    if (user_defined_config_file_name || fs::exists(conf_file_path)) {
+        parser.read(conf_file_path);
+    } else if (fs::exists(distribution_conf_file_path)) {
+        parser.read(distribution_conf_file_path);
+    }
+
+    // If the name of the configuration file was not specified, the files from the drop-in directories are also loaded.
+    // An alphabetically sorted list of all configuration files from the drop-in directories is created. If there is
+    // a file of the same name in both the user and distribution drop-in directories, only the user file is used.
```

Comment:
I am not sure about this logic. If user will specify -C /<path>, then all distribution and /etc overrides will be ignored. May I ask you for more details or some use case when this behavior is beneficial?

Reply:
@j-mracek 
`-C, --cacheonly`
You probably meant `-c [config file], --config [config file]`. Not yet implemented in dnf5. So `-c user_config_file`.
What should I do? The user has explicitly specified that they want configuration from this file. Overwriting it with configuration from drop-in directories is probably not what he wants. Therefore, in this case, the configuration from the drop-in directories is not loaded.
A compromise can still be made in this case. Change loading order. Load configuration files from drop-in directories and then from an explicitly user-defined file. But it will be so weird. The user changes the path to the main configuration file, but not the path to the drop-in directories.